### ISSUE TITLE
Question : regression tree ensemble predictions are inconsistent with the ones from microsoft/LightGBM

### ISSUE THREAD

**Issue by hufflepuff02** [Azure/mmlspark#868]
Hello,

I inherited from a stored tree model trained using the python API from microsoft/LightGBM, which from my understanding uses the same underlying C++ code as this one.

Since the saved .txt  file has the same format, I can load the model using the pyspark API provided in this repository. However the computed predictions are often slightly - sometimes completly - off when comparing to the ones I get using the original non-spark API.

This is a little odd to me as training aside, I would expect the evaluation of the same tree ensemble to give roughly identical results regardless of implementation. I'm guessing this stems from a different way of evaluating the model maybe due to distributed computing ?

While the interest of such a compatibility might be debatable, I would like to understand where lies the difference. Especially, if it is because of parallel computing, or if a certain set of parameters can help me achieve the same results.

Can you give me pointers so I can learn a little more about what is going on ?

Thanks !

**Comment by imatiach-msft** [comment]
maybe you are training on one C++ version but loading on a different one?  What is your lightgbm version?  What is the version you are using in mmlspark?  Different native code might account for the difference.  Recently I updated lightgbm dependency in mmlspark on master to use the very latest lightgbm (from their master), I wonder if that may cause your predictions to be more similar to what you are seeing on python API?

**Comment by hufflepuff02** [comment]
Hi @imatiach-msft 

Thank you for your answer.

I'm using the lastest releases, lightgbm 2.3.1 and mmlspark 1.0.0-rc1. I'm unsure what version of lightgbm mmlspark is running under the hood.

I'll try to use the latest versions from the masters.

**Comment by imatiach-msft** [comment]
@hufflepuff02  the latest version on mmlspark master is using a newer version of lightgbm (just updated a couple days ago), but I'm not sure if that will resolve your issue

**Comment by hufflepuff02** [comment]
@imatiach-msft Indeed, loading a jar built from mmlspark's master did not resolve the issue.

**Comment by imatiach-msft** [comment]
@hufflepuff02 sorry about the trouble you are having.  I do recall once another internal user had the same issue but when we went on a call we discovered the input data when calling python vs pyspark was slightly different (I think one of the columns they were pulling from SQL was different) which caused the slight difference in the predicted values.  That might not be the case for you however.  

1.) How large is the difference between the two in the prediction values?  
2.) Could you post corresponding python input data and pyspark input data and the output result?  

That may help diagnose the issue.

**Comment by hufflepuff02** [comment]
@imatiach-msft I'm afraid I can't share the data I'm working with.

However, I tried using another dataset (boston house pricing from sklearn datasets) and still get different results. I used the lastest releases again for convenience. Please see the notebook attached to this comment (I also included an html version in case it helps). Hopefully it will help figure out where my issue comes from.

[lgb_mml_example.zip](https://github.com/Azure/mmlspark/files/4730584/lgb_mml_example.zip)

Regarding the difference in prediction values, on 100 samples it ranges from ~12 to ~256000 with a mean of ~9000.

**Comment by imatiach-msft** [comment]
@hufflepuff02 sorry about the trouble you are having.  I spent quite a bit of time trying to repro, and would you believe it, when I ran the code I got the same result from mmlspark and lightgbm:

![image](https://user-images.githubusercontent.com/24683184/84106519-5a8ebf00-a9e9-11ea-805d-41a8db042d78.png)
 bizarre!

**Comment by imatiach-msft** [comment]
I tried the repro using lightgbm 2.3.1 and the latest version of mmlspark from master

**Comment by imatiach-msft** [comment]
@hufflepuff02 can you try to run the following code to see the full input passed to lightgbm:

**mml_model = LightGBMRegressionModel.loadNativeModelFromFile(model_path)
mml_predictions = mml_model.transform(dfin).withColumnRenamed('prediction', 'mml_prediction')
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', -1)
mml_predictions.toPandas()**

![image](https://user-images.githubusercontent.com/24683184/84107443-9dea2d00-a9eb-11ea-8355-3d0cb73931b9.png)
Just want to validate nothing strange is happening with the vector assembler.

Also, are you sure you are using latest master and not the 1.0 rc jar?  There have been a lot of fixes since then and I wonder if this was an issue that was either fixed since then or if we are missing corresponding values due to different lightgbm versions.

**Comment by imatiach-msft** [comment]
Sorry, one more thing, can you please send your OS and any other platform information?   I was testing on ubuntu (linux), I'm wondering if it's a platform specific issue?

**Comment by hufflepuff02** [comment]
@imatiach-msft yes I was still using the 1.0 rc jar. I tried using the lastest version from the master but I still get different results :
![Capture](https://user-images.githubusercontent.com/60149590/84151437-aa11d100-aa63-11ea-9c35-4bb39497fe60.PNG)

Regarding platform I'm working on ubuntu 16.04.6. I'm using spark 2.4.5 together with python 3.7.7 in a pipenv virtual environement.

**Comment by hufflepuff02** [comment]
I'm not sure what the "right" way to use the master would be. I couldn't get it using maven dependencies so I built jar and whl files from the master using `sbt packagePython`.

I added mmlspark to my pipenv using the whl, then I set up the spark conf like this :

```
spark = pyspark.sql.SparkSession.builder.appName('helga') \
    .config('spark.jars.repositories', 'https://mmlspark.azureedge.net/maven') \
    .config('spark.jars.packages', 'com.microsoft.ml.lightgbm:lightgbmlib:2.3.180') \
    .config('spark.jars', 'mmlspark_2.11-1.0.0-rc1-89-64481e94-SNAPSHOT.jar') \
    .getOrCreate()
```

**Comment by imatiach-msft** [comment]
@hufflepuff02 actually I realized I wasn't using master, I was using the rc version since that is what you are setting in your notebook (although I think the python code was master).  I was able to use the scala code from master too by modifying the app to be:

```
import pyspark
spark = pyspark.sql.SparkSession.builder.appName('hufflepower') \
    .config('spark.jars.packages', 'com.microsoft.ml.spark:mmlspark_2.11:1.0.0-rc1-88-8c00486f-SNAPSHOT') \
    .config('spark.jars.repositories', 'https://mmlspark.azureedge.net/maven') \
    .getOrCreate()

from pyspark.ml.feature import VectorAssembler
from mmlspark.lightgbm import LightGBMRegressionModel
```

The predictions were still the same though:

![image](https://user-images.githubusercontent.com/24683184/84169621-d0754380-aa46-11ea-8afa-26f10b66ee7b.png)

**Comment by imatiach-msft** [comment]
for building and using latest python on master, I just followed the guide:
https://github.com/Azure/mmlspark/blob/master/docs/developer-readme.md

I think these are the commands I ran (not tested, just copy-pasting from guide):
```
cd mmlspark
sbt setup
sbt compile
sbt createCondaEnv
sbt packagePython
sbt installPipPackage
```

but I don't think the python code should matter much.  Actually, this PR did change loadNativeModelFromFile recently, but I don't think it should matter:
https://github.com/Azure/mmlspark/pull/864/files#diff-d5072267cfdeb7ffaaff62256aeb3227L35

**Comment by imatiach-msft** [comment]
I tried running with a slightly earlier version of python code which does not include the change mentioned above (changeset 84b392c3a46cff8d2138326da960a912ed0baf75), but I am still not seeing the issue

**Comment by hufflepuff02** [comment]
@imatiach-msft I worked on reproducing the issue using a container environement. With my first attempt I got the right results, which is pretty good news.

I'd still like to understand the issue though, so I don't encounter it again in the future. I'll try to build a container that actually reproduces it and get back to you.
